<a href="https://colab.research.google.com/github/shwet369/Cosine-Similarity-Chatbot-main/blob/https%2Fgithub.com%2Fshwet369%2F-_Data-science/Document_search_using_GeminiAI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install -U -q google.generativeai

**Import Packages**

In [2]:
import google.generativeai as genai
import google.ai.generativelanguage as glm

**Get API key**

In [3]:
# Or use `os.getenv('API_KEY')` to fetch an environment variable.
API_KEY="AIzaSyAWEL056CjzC177E5UQ2MYEl3xNhL0bhFY"

genai.configure(api_key=API_KEY)

**Get the embedd model**

In [4]:
for m in genai.list_models():
  if 'embedContent' in m.supported_generation_methods:
    print(m.name)

models/embedding-001
models/text-embedding-004


In [5]:
sample_text = ("Title: The next generation of AI for developers and Google Workspace"
    "\n"
    "Full article:\n"
    "\n"
    "Gemini API & Google AI Studio: An approachable way to explore and prototype with generative AI applications")
print(sample_text)

Title: The next generation of AI for developers and Google Workspace
Full article:

Gemini API & Google AI Studio: An approachable way to explore and prototype with generative AI applications


**Get the embeddings for sample text**

In [6]:
title = "The next generation of AI for developers and Google Workspace"
sample_text = ("Title: The next generation of AI for developers and Google Workspace"
    "\n"
    "Full article:\n"
    "\n"
    "Gemini API & Google AI Studio: An approachable way to explore and prototype with generative AI applications")

model = 'models/embedding-001'
embedding = genai.embed_content(model=model,
                                content=sample_text,
                                task_type="retrieval_document",
                                title=title)

print(embedding)

{'embedding': [0.034113422, -0.055176623, -0.020209068, -0.0041249595, 0.058917794, 0.01412951, 0.004535358, 0.0014303708, 0.059766337, 0.08292115, 0.007162965, 0.0069041667, -0.05308343, -0.01090513, 0.03214021, -0.037164, 0.050372463, 0.019348344, -0.037328612, 0.026647933, 0.03078176, -0.011288503, -0.031485256, -0.060248997, -0.026219437, -0.009794238, 0.0066301385, -0.018465156, -0.026324723, 0.020442627, -0.06317685, 0.014559578, -0.052296046, 0.016451124, -9.71961e-05, -0.051706687, -0.0054406007, -0.05696762, 0.01114414, -0.009201795, -0.0021951075, -0.10997012, -0.0117121935, 0.021221716, 0.009171805, -0.029621972, 0.034534886, 0.039578073, 0.019021517, -0.062691696, 0.039473332, 0.052403256, 0.061814193, -0.034507953, -0.009557814, -0.004955104, 0.01783901, -0.021176832, 0.015043591, 0.015390573, -0.0063342857, 0.043696415, -0.028341988, 0.02843402, 0.014726859, -0.06585564, -0.044533547, 0.0055523184, 0.035775978, 0.031099156, 0.027357664, 0.028062243, 0.056972917, -0.054656

In [7]:
len(embedding['embedding'])

768

**Get the Embeddings for Documents**

In [9]:
DOCUMENT1 = {
    "title": "Operating the Climate Control System",
    "content": "Your Googlecar has a climate control system that allows you to adjust the temperature and airflow in the car. To operate the climate control system, use the buttons and knobs located on the center console.  Temperature: The temperature knob controls the temperature inside the car. Turn the knob clockwise to increase the temperature or counterclockwise to decrease the temperature. Airflow: The airflow knob controls the amount of airflow inside the car. Turn the knob clockwise to increase the airflow or counterclockwise to decrease the airflow. Fan speed: The fan speed knob controls the speed of the fan. Turn the knob clockwise to increase the fan speed or counterclockwise to decrease the fan speed. Mode: The mode button allows you to select the desired mode. The available modes are: Auto: The car will automatically adjust the temperature and airflow to maintain a comfortable level. Cool: The car will blow cool air into the car. Heat: The car will blow warm air into the car. Defrost: The car will blow warm air onto the windshield to defrost it."}
DOCUMENT2 = {
    "title": "Touchscreen",
    "content": "Your Googlecar has a large touchscreen display that provides access to a variety of features, including navigation, entertainment, and climate control. To use the touchscreen display, simply touch the desired icon.  For example, you can touch the \"Navigation\" icon to get directions to your destination or touch the \"Music\" icon to play your favorite songs."}
DOCUMENT3 = {
    "title": "Shifting Gears",
    "content": "Your Googlecar has an automatic transmission. To shift gears, simply move the shift lever to the desired position.  Park: This position is used when you are parked. The wheels are locked and the car cannot move. Reverse: This position is used to back up. Neutral: This position is used when you are stopped at a light or in traffic. The car is not in gear and will not move unless you press the gas pedal. Drive: This position is used to drive forward. Low: This position is used for driving in snow or other slippery conditions."}

documents = [DOCUMENT1, DOCUMENT2, DOCUMENT3]

- There are 3 documents availabel

- Each document will pass through the embedding model

- Embedding model will give  a vector with 768 values

- V1  V2   V3

- Now user ask a question

- Qn also pass to the embedd vector then convert into vector

- VQ


- Vq dot product with v1 : Similarity value1

- Vq dot product with v2 : SV2

- Vq dot product with v3 : SV3



In [10]:
import pandas as pd
df = pd.DataFrame(documents)
df.columns = ['Title', 'Text']
df

Title  \
0  Operating the Climate Control System   
1                           Touchscreen   
2                        Shifting Gears   

                                                Text  
0  Your Googlecar has a climate control system th...  
1  Your Googlecar has a large touchscreen display...  
2  Your Googlecar has an automatic transmission. ...

In [11]:
# Get the embeddings of each text and add to an embeddings column in the dataframe
def embed_fn(title, text):
  return genai.embed_content(model=model,
                             content=text,
                             task_type="retrieval_document",
                             title=title)["embedding"]

df['Embeddings'] = df.apply(lambda row: embed_fn(row['Title'], row['Text']), axis=1)
df

Title  \
0  Operating the Climate Control System   
1                           Touchscreen   
2                        Shifting Gears   

                                                Text  \
0  Your Googlecar has a climate control system th...   
1  Your Googlecar has a large touchscreen display...   
2  Your Googlecar has an automatic transmission. ...   

                                          Embeddings  
0  [-0.03336111, -0.021217091, -0.04958193, -0.01...  
1  [0.009660737, -0.030662702, -0.01728142, -0.05...  
2  [-0.042707954, -0.007160867, -0.03242516, -0.0...

In [12]:
v1=genai.embed_content(model=model,
                             content=DOCUMENT1['content'],
                             task_type="retrieval_document",
                             title=DOCUMENT1['title'])["embedding"]
v2=genai.embed_content(model=model,
                             content=DOCUMENT2['content'],
                             task_type="retrieval_document",
                             title=DOCUMENT2['title'])["embedding"]
v3=genai.embed_content(model=model,
                             content=DOCUMENT3['content'],
                             task_type="retrieval_document",
                             title=DOCUMENT3['title'])["embedding"]

In [13]:
v=[v1,v2,v3]
df['Vectors']=v
df

Title  \
0  Operating the Climate Control System   
1                           Touchscreen   
2                        Shifting Gears   

                                                Text  \
0  Your Googlecar has a climate control system th...   
1  Your Googlecar has a large touchscreen display...   
2  Your Googlecar has an automatic transmission. ...   

                                          Embeddings  \
0  [-0.03336111, -0.021217091, -0.04958193, -0.01...   
1  [0.009660737, -0.030662702, -0.01728142, -0.05...   
2  [-0.042707954, -0.007160867, -0.03242516, -0.0...   

                                             Vectors  
0  [-0.03336111, -0.021217091, -0.04958193, -0.01...  
1  [0.009660737, -0.030662702, -0.01728142, -0.05...  
2  [-0.042707954, -0.007160867, -0.03242516, -0.0...

**Retrival query**

- When ever if you want ask a qn we need to pass text only

- So the task type is **retrival query**

In [14]:
query = "How do you shift gears in the Google car?"
model = 'models/embedding-001'

request = genai.embed_content(model=model,
                              content=query,
                              task_type="retrieval_query")

qv=request['embedding']

# genai.embed_content(model=model,
#                              content=DOCUMENT1['content'],
#                              task_type="retrieval_document",
#                              title=DOCUMENT1['title'])

In [15]:
import numpy as np
np.dot(v1,qv),np.dot(v2,qv),np.dot(v3,qv)

(0.6679797316031354, 0.6614951055467841, 0.7768085069703565)

**Get the similairity**

In [16]:
# Call embedding vectors
df['Embeddings']

def similarity(document_vector,query_vector):
    return(np.dot(document_vector,query_vector))

list1=[similarity(i,qv) for i in df['Embeddings']]
print(list1)
max_index=np.argmax(list1)
print(max_index)
documents[np.argmax(list1)]

[0.6679797316031354, 0.6614951055467841, 0.7768085069703565]
2


{'title': 'Shifting Gears',
 'content': 'Your Googlecar has an automatic transmission. To shift gears, simply move the shift lever to the desired position.  Park: This position is used when you are parked. The wheels are locked and the car cannot move. Reverse: This position is used to back up. Neutral: This position is used when you are stopped at a light or in traffic. The car is not in gear and will not move unless you press the gas pedal. Drive: This position is used to drive forward. Low: This position is used for driving in snow or other slippery conditions.'}

In [17]:
def similarity(document_vector,query_vector):
    return(np.dot(document_vector,query_vector))
list1=[similarity(i,qv) for i in v]
documents[np.argmax(list1)]

{'title': 'Shifting Gears',
 'content': 'Your Googlecar has an automatic transmission. To shift gears, simply move the shift lever to the desired position.  Park: This position is used when you are parked. The wheels are locked and the car cannot move. Reverse: This position is used to back up. Neutral: This position is used when you are stopped at a light or in traffic. The car is not in gear and will not move unless you press the gas pedal. Drive: This position is used to drive forward. Low: This position is used for driving in snow or other slippery conditions.'}

In [18]:
DOCUMENT3

{'title': 'Shifting Gears',
 'content': 'Your Googlecar has an automatic transmission. To shift gears, simply move the shift lever to the desired position.  Park: This position is used when you are parked. The wheels are locked and the car cannot move. Reverse: This position is used to back up. Neutral: This position is used when you are stopped at a light or in traffic. The car is not in gear and will not move unless you press the gas pedal. Drive: This position is used to drive forward. Low: This position is used for driving in snow or other slippery conditions.'}

In [19]:
query = "How google car touch screen works?"
model = 'models/embedding-001'

request = genai.embed_content(model=model,
                              content=query,
                              task_type="retrieval_query")

qv=request['embedding']


def similarity(document_vector,query_vector):
    return(np.dot(document_vector,query_vector))
list1=[similarity(i,qv) for i in v]
documents[np.argmax(list1)]

{'title': 'Touchscreen',
 'content': 'Your Googlecar has a large touchscreen display that provides access to a variety of features, including navigation, entertainment, and climate control. To use the touchscreen display, simply touch the desired icon.  For example, you can touch the "Navigation" icon to get directions to your destination or touch the "Music" icon to play your favorite songs.'}

## TASK

- Read some documents from google

- Phy  Chem  Bio

- Create a format of documents

- Perform end to end

- Show me the output on suitable question